src/components/Sheet/Sheet.tsx

```diff
@@ -0,0 +1,180 @@
+import * as React from 'react';
+
+import {CSSTransition} from 'react-transition-group';
+import debounce from 'lodash/debounce';
+import {classNames} from '@shopify/react-utilities/styles';
+
+import {navigationBarCollapsed} from '../../utilities/breakpoints';
+import {Key} from '../../types';
+import {layer, overlay, Duration} from '../shared';
+import {frameContextTypes, FrameContext} from '../Frame';
+import {withAppProvider, WithAppProviderProps} from '../AppProvider';
+
+import Backdrop from '../Backdrop';
+import TrapFocus from '../TrapFocus';
+import Portal from '../Portal';
+import KeypressListener from '../KeypressListener';
+import EventListener from '../EventListener';
+
+import styles from './Sheet.scss';
+
+export const BOTTOM_CLASS_NAMES = {
+  enter: classNames(styles.Bottom, styles.enterBottom),
+  enterActive: classNames(styles.Bottom, styles.enterBottomActive),
+  exit: classNames(styles.Bottom, styles.exitBottom),
+  exitActive: classNames(styles.Bottom, styles.exitBottomActive),
+};
+
+export const RIGHT_CLASS_NAMES = {
+  enter: classNames(styles.Right, styles.enterRight),
+  enterActive: classNames(styles.Right, styles.enterRightActive),
+  exit: classNames(styles.Right, styles.exitRight),
+  exitActive: classNames(styles.Right, styles.exitRightActive),
+};
+
+export interface Props {
+  /** Whether or not the sheet is open */
+  open: boolean;
+  /** The child elements to render in the sheet */
+  children: React.ReactNode;
+  /** Callback when the backdrop is clicked or `ESC` is pressed */
+  onClose(): void;
+}
+
+export type ComposedProps = Props & WithAppProviderProps;
+
+export interface State {
+  mobile: boolean;
+}
+
+export class Sheet extends React.Component<ComposedProps, State> {
+  static contextTypes = frameContextTypes;
+  context: FrameContext;
+
+  state: State = {
+    mobile: false,
+  };
+
+  private handleResize = debounce(
+    () => {
+      const {
+        state: {mobile},
+        handleToggleMobile,
+      } = this;
+
+      if (mobile !== isMobile()) {
+        handleToggleMobile();
+      }
+    },
+    40,
+    {leading: true, trailing: true, maxWait: 40},
+  );
+
+  componentDidMount() {
+    const {
+      state: {mobile},
+      context: {frame},
+      props: {
+        polaris: {intl},
+      },
+      handleToggleMobile,
+    } = this;
+
+    if (frame == null) {
+      // eslint-disable-next-line no-console
+      console.warn(intl.translate('Polaris.Sheet.warningMessage'));
+    }
+
+    if (mobile !== isMobile()) {
```

Comment:
I don't think so, since we want to avoid calling `handleToggleMobile` each time this line of code is reached. The idea here is if the component's mobile state doesn't equate to the viewport's mobile state, we want to run the toggle